### ISSUE TITLE
feature suggestion - generate a CLI for a model

### ISSUE THREAD

**Issue by guymguym** [awslabs/smithy-rs#1362]
What I wanted was to generate a [clap](https://github.com/clap-rs/clap) command for every API operation to make a CLI for any smithy model.
I did a partial implementation for my case (did not handle all types) for the s3 API that you can see [here](https://github.com/guymguym/s3d/blob/e396d106f386b881b733b0ebd3697ed3789d582c/codegen/gen_commands.rs
).

I think this might be useful in general.
Basically every operation can generate a separate command that uses a generated client to call the api.
Docs are already populated in the CLI from the docs in the model.
Inputs are read from flags. Need to handle complex types somehow (I guess json is most natural).
Payload is read from stdin or file.

What do you think? Would that be useful to other projects? 
I would be happy to suggest a PR if you think this fits the project.

**Comment by crisidev** [comment]
I think it is an interesting idea which could help a lot with debugging / troubleshooting. AFAIK it is not something we have in our roadmap currently, but contributions are always welcome :)

Let's see what the client SDK folks have to say about this.

**Comment by rcoh** [comment]
yeah this has come up before—I think this could be an awesome use of the existing customization machinery—I suspect it would not even be that hard, just not something on our roadmap.

I'm not positive we'd want to host it within the smithy-rs repo, at least not right away. However, you could definitely to it in a separate repo, using the smithy-rs repo as a gradle composite build or git submodule

**Comment by crisidev** [comment]
To be honest doing it in a separate repo will be a pain since you need the submodule :(

**Comment by guymguym** [comment]
@rcoh Are there reasons not to include this functionality in this repo? I think it wouldn't be easily maintainable elsewhere with all the semantic changes coming from smithy itself.

**Comment by rcoh** [comment]
Mostly just minimalism—if we add something to this repo we're expressing a commitment to maintain it going forward and a CLI currently isn't in our mandate / roadmap

**Comment by hiljusti** [comment]
[Start with no](https://basecamp.com/gettingreal/05.3-start-with-no) is reasonable.

I'd really, _really_ like to be able to generate CLIs from a Smithy model someday. I can talk through the potential value to Amazon/AWS itself privately if it's helpful. For a tl;dr: I believe a CLI should be on the roadmap, even if it's not a top-ten priority.

A CLI is valuable but non-trivial. There's a lot of ambiguity to sift through, boundaries to define, and steep [Hyrum's Law](https://www.hyrumslaw.com/) consequences of any choice along the way.
